sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,243 @@
+} dif_spi_device_irq_state_t;
+
+/**
+ * Represents the types of interrupts that the SPI device will
+ * fire to signal state.
+ */
+typedef enum dif_spi_device_irq_type {
+  kDifSpiDeviceIrqTypeRxFull,
+  kDifSpiDeviceIrqTypeRxAboveLevel,
+  kDifSpiDeviceIrqTypeTxBelowLevel,
+  kDifSpiDeviceIrqTypeRxError,
+  kDifSpiDeviceIrqTypeRxOverflow,
+  kDifSpiDeviceIrqTypeTxUnderflow,
+} dif_spi_device_irq_type_t;
+
+/**
+ * Errors returned by SPI DIF functions.
+ */
+typedef enum dif_spi_device_result {
+  /**
+   * Indicates a successful operation.
+   */
+  kDifSpiDeviceResultOk = 0,
+  /**
+   * Indicates a failed precondition on the function arguments.
+   */
+  kDifSpiDeviceResultBadArg,
+} dif_spi_device_result_t;
+
+/**
+ * The length of the SPI device SRAM, in bytes.
+ *
+ * Useful for initializing FIFO lengths: for example, for equally-sized FIFOs,
+ * `rx_fifo_len` and `tx_fifo_len` would be set to `kDifSpiDeviceSramLen / 2`.
+ */
+extern const uint16_t kDifSpiDeviceSramLen;
+
+/**
+ * Configuration for initializing a SPI device.
+ */
+typedef struct dif_spi_device_config {
+  dif_spi_device_edge_t clock_polarity;
+  dif_spi_device_edge_t data_phase;
+  dif_spi_device_bit_order_t tx_order;
+  dif_spi_device_bit_order_t rx_order;
+  uint8_t rx_fifo_timeout;
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_len;
+} dif_spi_device_config_t;
+
+/**
+ * State for a particular SPI device.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_spi_device {
+  mmio_region_t base_addr;
+  uint16_t rx_fifo_base;
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_base;
+  uint16_t tx_fifo_len;
+} dif_spi_device_t;
+
+/**
+ * Initializes a SPI device with the given configuration.
+ *
+ * @param base_addr The start of the SPI device register.
+ * @param config The configuration to initialize with.
+ * @param spi Out param for the initialized device.
+ * @return The result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_init(
+    mmio_region_t base_addr, const dif_spi_device_config_t *config,
+    dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ * halt.
+ *
+ * @param spi A SPI device.
+ * @return The result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi);
+
+/**
+ * Resets all interrupt-related state on the given SPI device, such as enabled
+ * interrupts and set RX/TX levels.
+ *
+ * @param spi A SPI device.
+ * @return The result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_reset(const dif_spi_device_t *spi);
+
+/**
+ * Checks whether a particular interrupt has been enabled or disabled.
+ *
+ * @param spi A SPI device.
+ * @param type Which IRQ type to check.
+ * @param flag Out param for the enablement state.
+ * @return The result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_get(
```

Comment:
Note for myself: this function's doc is actually completely wrong... this tests for whether an interrupt is currently being *serviced*. I'll update that before merging.